### PR TITLE
enhancement: add tooltip text to sidebar icons

### PR DESCRIPTION
**Pull request by tarunsamanta2k20**
Add tooltip text to the sidebar icons so that users can understand the icons when the sidebar is collapsed.

closes: #392 

image

### CODE REVIEW

**Comment by tarunsamanta2k20** [comment]
I haven't handled the is expanded case yet.

File: src/components/Sidebar/SidebarIcon.tsx [hunk 1]
```diff
@@ -11,38 +11,39 @@ export interface ISidebarIconProps {
...
-  onClick,
-  children,
-}: ISidebarIconProps) => {
-  return (
-    <button
-      type="button"
-      onClick={() => onClick()}
-      className={`sidebar-background-color-transition flex flex-row align-middle w-full rounded-lg
+const SidebarIcon = forwardRef<HTMLButtonElement, ISidebarIconProps>(
+  function SidebarIcon(props, ref) {
```
Comment (ajmcquilkin) [inline src/components/Sidebar/SidebarIcon.tsx:15]: Was there a reason that you chose to keep `DefaultTooltip` outside of the `SidebarIcon` component and instead forwarded the ref in?

Comment (tarunsamanta2k20) [reply src/components/Sidebar/SidebarIcon.tsx:15]: @

> Was there a reason that you chose to keep `DefaultTooltip` outside of the `SidebarIcon` component and instead forwarded the ref in?

@ajmcquilkin it won't show anything if I don't use forward ref or other HTML element as Sidebar Icon is a custom component.

Comment (tarunsamanta2k20) [reply src/components/Sidebar/SidebarIcon.tsx:15]: @ajmcquilkin I know this syntax.

Comment (ajmcquilkin) [reply src/components/Sidebar/SidebarIcon.tsx:15]: Sorry, I should have phrased that better. Instead of writing:

```typescript
<SidebarTab title="Overview" isSidebarExpanded={isSidebarExpanded}>
  <DefaultTooltip text="Messaging">
    <SidebarIcon
      name="Messaging"
      isActive={location.pathname === AppRoutes.MESSAGING}
      isSidebarExpanded={isSidebarExpanded}
      onClick={() => navigateTo(AppRoutes.MESSAGING)}
    >
      <MessagesSquare strokeWidth={1.5} className="w-6 h-6" />
    </SidebarIcon>
  </DefaultTooltip>
</SidebarTab>
```

Is there a reason you didn't define `SidebarIcon` as:

```typescript
<DefaultTooltip ... >
  <button ... >
     ... contents of SidebarIcon ...
  </button>
</SidebarIcon>
```

I think it would be fine forcing all instances of `SidebarIcon` to have a tooltip, and this would make disabling the tooltip easier with the `isActive` prop

**Comment by ajmcquilkin** [comment]
Thanks for the PR, overall change looks good! Just have the comment on the ref forwarding, and a small comment on switching the tooltip for when expanding the sidebar. This could probably be a quick ternary.

image

File: src/components/Sidebar/SidebarIcon.tsx (hunk 1 above)
Comment (tarunsamanta2k20) [reply src/components/Sidebar/SidebarIcon.tsx:15]: > Sorry, I should have phrased that better. Instead of writing:
> 
> ```ts
> <SidebarTab title="Overview" isSidebarExpanded={isSidebarExpanded}>
>   <DefaultTooltip text="Messaging">
>     <SidebarIcon
>       name="Messaging"
>       isActive={location.pathname === AppRoutes.MESSAGING}
>       isSidebarExpanded={isSidebarExpanded}
>       onClick={() => navigateTo(AppRoutes.MESSAGING)}
>     >
>       <MessagesSquare strokeWidth={1.5} className="w-6 h-6" />
>     </SidebarIcon>
>   </DefaultTooltip>
> </SidebarTab>
> ```
> 
> Is there a reason you didn't define `SidebarIcon` as:
> 
> ```ts
> <DefaultTooltip ... >
>   <button ... >
>      ... contents of SidebarIcon ...
>   </button>
> </SidebarIcon>
> ```
> 
> I think it would be fine forcing all instances of `SidebarIcon` to have a tooltip, and this would make disabling the tooltip easier with the `isActive` prop

@ajmcquilkin you are like that?

<DefaultTooltip text="Messaging">
              <button>
              <SidebarIcon
                name="Messaging"
                isActive={location.pathname === AppRoutes.MESSAGING}
                isSidebarExpanded={isSidebarExpanded}
                onClick={() => navigateTo(AppRoutes.MESSAGING)}
              >
                <MessagesSquare strokeWidth={1.5} className="w-6 h-6" />
              </SidebarIcon>
              </button>
            </DefaultTooltip>

Comment (ajmcquilkin) [reply src/components/Sidebar/SidebarIcon.tsx:15]: Sorry, not sure I'm understanding your comment. My question is just out of curiosity, doesn't seem to make a functionality difference.

**Comment by tarunsamanta2k20** [comment]
@ajmcquilkin let me modify all those stuffs.

**Comment by tarunsamanta2k20** [comment]
@ajmcquilkin a small ask. where do you want to handle the is Expanded condition?

**Comment by ajmcquilkin** [comment]
Oh the condition to disable the tooltip when the sidebar icon is expanded? I was thinking a bit more about that, it might be worth passing a new `deactivated` prop to the `DefaultTooltip` component, then include the following check:

```typescript
if (deactivate) return <>{children}</>;

return ( ... existing UI code ...)
```

That would mean you wouldn't have to include any ugly ternaries in the `SidebarIcon` component

**Comment by ajmcquilkin** [comment]
Thanks for the changes, let me check them out locally!

**Comment by ajmcquilkin** [comment]
Hmm interesting, seems that caused a styling issue. 

image

image

A couple of thoughts:

- Probably want to make the `deactivated` prop optional defaulting to `false`, we're getting build errors from the other places `DefaultTooltip` is used
- I'm wondering if the styling issue is caused by some width CSS issue because the `SidebarIcon` is being wrapped. Maybe try something like this?

```typescript
const SidebarIcon = ({
  name,
  isActive,
  isSidebarExpanded,
  tooltipText,
  onClick,
  children,
}: ISidebarIconProps) => {
  return (
    <DefaultTooltip text={tooltipText} deactivated={false}>
      <button
        type="button"
        ... more props ...
      >
        ... existing sidebar icon content ...
      </button>
    </DefaultTooltip>
  );
};
```

In that case, we wouldn't need to use `DefaultTooltip` within the `Sidebar` component directly and would only need to add the `tooltipText` prop to each instance of `SidebarIcon`

**Comment by tarunsamanta2k20** [comment]
@ajmcquilkin I didn't get it for styling issue.

**Comment by ajmcquilkin** [comment]
Sorry, let me rephrase. I'm suggesting moving where we call the `DefaultTooltip` component, from the `Sidebar` component to the `SidebarIcon` component.

The question I'm thinking about is whether there's ever a case in which you would want a `SidebarIcon` to not have a tooltip. I would argue that we want all `SidebarIcon` components to have tooltips. I would imagine we would then get something looking like the following code:

_Sidebar.tsx_
```typescript
const Sidebar = () => {
  const [isSidebarExpanded, setSidebarExpanded] = useState(true);

  const navigateTo = useNavigate();
  const location = useLocation();

  return (
    <div
      className={`shrink-0 sidebar-width-transition flex flex-col h-screen box-content shadow-lg overflow-y-auto hide-scrollbar border-r border-gray-100 overflow-x-hidden ${
        isSidebarExpanded ? "w-[300px] text-gray-500" : "w-20 text-white"
      }`}
    >
      <SidebarLogo isSidebarExpanded={isSidebarExpanded} />

      <div className="flex flex-col flex-1 justify-between px-4 pt-4 pb-1">
        <div className="flex flex-col gap-6 mb-6">
          <SidebarTab title="Overview" isSidebarExpanded={isSidebarExpanded}>
              ... other icons ...

              <SidebarIcon
                name="Configure Radio"
                isActive={location.pathname.includes(AppRoutes.CONFIGURE_RADIO)}
                isSidebarExpanded={isSidebarExpanded}
                tooltipText="Configure Radio"
                onClick={() => navigateTo(AppRoutes.CONFIGURE_RADIO)}
              >
                <Router strokeWidth={1.5} className="w-6 h-6" />
              </SidebarIcon>

              <SidebarIcon
                name="Configure Modules"
                isActive={location.pathname.includes(
                  AppRoutes.CONFIGURE_PLUGINS
                )}
                isSidebarExpanded={isSidebarExpanded}
                tooltipText="Configure Modules"
                onClick={() => navigateTo(AppRoutes.CONFIGURE_PLUGINS)}
              >
                <Component strokeWidth={1.5} className="w-6 h-6" />
              </SidebarIcon>

               ... other icons ...
          </SidebarTab>
        </div>
      </div>
    </div>
  );
};
```

_SidebarIcon.tsx_
```typescript
const SidebarIcon = ({
  name,
  isActive,
  isSidebarExpanded,
  tooltipText,
  onClick,
  children,
}: ISidebarIconProps) => {
  return (
    <DefaultTooltip text={tooltipText} deactivated={false}>
      <button
        type="button"
        ... more props ...
      >
        ... existing sidebar icon content ...
      </button>
    </DefaultTooltip>
  );
};
```

**Comment by tarunsamanta2k20** [comment]
@ajmcquilkin is it working as expected?

**Comment by ajmcquilkin** [comment]
Looks great, thank you so much for the contribution! Approving

**Review by ajmcquilkin — APPROVED** [review]

**Comment by tarunsamanta2k20** [comment]
@ajmcquilkin is all done for this PR?

**Comment by ajmcquilkin** [comment]
Yep, all good for merge! Sorry, I should have clarified. Merging now.
